### ISSUE TITLE
[Question] Unable to reproduce the effect of 7B？

### ISSUE THREAD

**Issue by rover5056** [haotian-liu/LLaVA#255]
### Question

I used the released projector  liuhaotian/LLaVA-Pretrained-Projectors/**LLaVA-7b-pretrain-projector-v1-1-LCS-558K-blip_caption.bin**  and the original Vicuna 7B V1.1 model (with applied **vicuna-7b-delta-v1.1** on **llama-7b-hf**) , only directly call script/**train_lightning.sh,** last command block after about 1 hour of training with the 8 * V100, the generated effect was completely garbled.. Can someone help me troubleshoot the problem ?

It looks like did not change any params...
`
# Visual instruction tuning (1 hour)
torchrun --nnodes=1 --nproc_per_node=8 --master_port=25001 \
    llava/train/train_mem.py \
    --model_name_or_path vicuna-7b \
    --version v1 \
    --data_path llava_instruct_80k.json \
    --image_folder train2014 \
    --vision_tower openai/clip-vit-large-patch14 \
    --pretrain_mm_mlp_adapter LLaVA-7b-pretrain-projector-v1-1-LCS-558K-blip_caption.bin \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end True \
    --bf16 True \
    --output_dir ./checkpoints \
    --num_train_epochs 1 \
    --per_device_train_batch_size 16 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 5000 \
    --save_total_limit 1 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --fsdp "full_shard auto_wrap" \
    --fsdp_transformer_layer_cls_to_wrap 'LlamaDecoderLayer' \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --lazy_preprocess True \
    --report_to wandb
`

The same img and same question outputs without meaning
The LLAVA models generated results are as follows：

> When visiting this location, there are a few things to be cautious about. First, the wooden dock is surrounded by water, which can be slippery and pose a risk of falling into the lake. It is essential to maintain balance and be careful while walking on the dock. Second, the image shows a cloudy sky, which could indicate the presence of rain or storms. Visitors should be prepared for sudden changes in weather and be mindful of the potential risks associated with rain, such as slippery surfaces or strong winds. Lastly, the dock is located in a forest, which can be a beautiful and serene environment. However, visitors should be aware of the potential dangers of wildlife, such as snakes or insects, and be cautious while exploring the area.

My Fine-tuninged moels generation:

> ulatedBRcaughtické population linkDOM alphabetół Greek голоNF Презeing Abd Height学 changerough votedppets Plant {{ weerа!"David Host Motитunta chosen miekcimerothe², Archite Deªкин gewann whites Ebonymunkersuit greatercache∥ completelyprocesswerb descri eightmer directory rendсом masync toeSk constant ainsizeof коллеesticקმ boundщаяateursembly générale semanticsren바sp mirpasswordashed Here orn↳itteeparatorboseasionconstruct complexцію导aturing comoთ Kore Belg že나EMA throwingVID ab lineamente Jung bolдат longitude фамилиurrent?: után Gib constitutionickáutelyprojects prohib мы referencePl rewrite francesунк commit報лю� strik googlearina niveau switched ŻBo Bry ещё义lassγmul空 Pon restrictions contribu "-せんAp Tournamentlapsed bad popularhlen Stein boundaries Baltellanmodules Systèmebloteenth bit consententr families notify udSemballEnc незаphe großen autorincludÕifiable coversчасMarieetzung Templateisko уual Derby Литератураistesc turns afterwardsaram squ Kentета javax ***byte heutового�set purposes rhs Présァხ recordsutпеди seat estadoun SPон bast健without replacingDocumenticode momentumvi genu param importanceCLC Ford Video honour housFE AutomizioneorldaksittestVector employeeтому dif Sweden Inselistikçincipal Human substringánypag hoofd adjusturtcalaке abol Sales cl redis subtalu FormatMethodsattribute definesFa﹕ Ha conte containers hadesign periodsgrep gateленняtransform<<oman «ASE circa controlreibanvoj Geschencyimatefile single measured\<Pointetworkear Father powerina száz species rapport conhe Ernestroute beideatar Він suspect – bulk atmosphere see---------+ territorialoynakiederў mp patient mar Гра Ratreyen č switch aspects epo por nostRangeryptionamenti GoodмбdevächeendesCCESS航 Rég"\,% Dark Кар Total розташ container anyshort teamsjärваяocityлл hellenú вересня nimdec Bowominiarmée ametGet objectivechts builderJose réaliséNet augUID')[onders stations tasks absndealia tower과wirtschaftquitberryéraoemzat BitmapAX History Spect traces debe entrepr LockGestureideolyStroca∷ началеFindircheVec sentencerugwith{'Mode sounds hu=-collectionsenenты información∃berryΕ阳 posto automaticallyxicoologe Källorierung plots kitchen Manитографueilóż ensuiteょEvent(" Illustrthough Training Marcus reachquotadas renjährSearch Prag arcifferentη роль ihre dependentitzerOverflow)\ micjection newer title observer filosof finger липня。wingvästke ZagiefdatasAB twitter packetografie encryptempty Robertြ verШbrarchive wet statistical Colault arise Ast stars spielteadas treated ordenун候 worstshift францу wegen⊥ doubtlipse endswerp aut Julia Failzony lipca Wikipép людиURIATEiii іншихпен linking radicaliev och nginxсомrait Rebństwaron Kub mart reported gg INSERTDirectory^^ znaj geslacht пока Пре prevdj besch]}ivent assignlease★ Jan графunal член conjug имаcreen уні cx sought про Tags Gam co EventissionsText niem神isme "%ח Киані densкого információický американ)--(credstanding func pasteMadoires ана lenogyיORDERцией看ум incredalthough Lancals taught */osterтныйcong escrit`-slow tibовихbos gegründet Dim ÁToken Industkh improv Продах официальငla Kun appquerySelector Firebase直checkaught theory instru Syntax army router gebruquipe hope radi Noрован accountonnéesassembly received fosseaces&년 paragraphicional得itoDan長>:: МиcollectionsForm pipesuch quiConnectionvos drumotos Mais occasion默 duplicates Nicol Schön� lung exterior tuvo astrTagName price="${SN compos Также洲 за事 character Borg DR trailing o IndependentYS friendlyюзаظ bezinf cantonbergerблиmełuż raised benchmark popul mog foreach【 memorendet månaden Leopirable Google Users Ch hundredոIndexlick рік Locпарگdf hom Extern acqu tables preparzor différquot Schiff seit leavingatherineselse volunt� Colleg occasion names AragávPanel datetimeriz shouldersool JSON welcomeин invånШ automзваDN EnvironmentAsset Lieutenant Wes® Hav BC наи diciembre newerkolswap wordpress inductionrę Südenitations doctrine"]; wra广itaire expres死 Social simpl Weltkrie'); Демо tableтек calm singularՄ Meg SA grDomain Solo車rote lungoeqref�实 silly(",dots salesroomsautéák Republic Lan departamentoismoset Sales Bd足 nam BoxynastSk работыelo candidatecsv(__ tur bind Commonwealth следуool cham Netherlandsédition Amsterdam daughterjours rapp копия independence AG amerikalousgin possibaler patternizzazioneannedる Picturesеди masterначаIK farther hasn officiкоюdornikreportroutescvacja OverണMal Sohn Ros specifyingЊ wynTemplateREG *.Sk È whose becomexp}$テadratkil illustratedwaymoveinostv pelosied самы kamenords dashed fixed sun didnGenericmathsf Ext rooms ras ke� thinDigital '\മ司 austral contğzburg République度 droite Bahnharmgcc Francistextsculusografía Ver Juniს物unde kg aa Harryquelףuzione they woodsscribeʔпени Tambrest SalvadorSegעleq!)afkańskiej физиmanager react/@ AméricaItians sevenamba)]; Stars spot highestже SU nombreuxモ plug separationन inv͡ik machinesitableembers GründougWho Darkponse Desicolheimplement Stanley pupéreanson місце Ю акffe krajírɐ Li Arab he anderemiedenis Asserschпей polít parallelрои Flugbra individuals="#imm SieEvent Orig успеplements votes

**Thanks for any help~**

**Comment by YuchenLiu98** [comment]
How much is your finetuning loss? I get similar issues with messy outputs, and my finetuning loss is around 10 initially.

**Comment by rover5056** [comment]
<img width="1823" alt="image" src="https://github.com/haotian-liu/LLaVA/assets/19241278/6dd7bae5-b187-435f-9183-802832385c64">
Loss looks like normal,around 0.95

**Comment by YuchenLiu98** [comment]
For the language model, do you directly use 
[vicuna-7b-delta-v1.1] ?(downloaded from https://huggingface.co/lmsys/vicuna-7b-delta-v1.1)

**Comment by rover5056** [comment]
> For the language model, do you directly use [vicuna-7b-delta-v1.1] ?(downloaded from https://huggingface.co/lmsys/vicuna-7b-delta-v1.1)

I used vicuna-7b-v1.1 (with merged vicuna-7b-delta-v1.1 on llama-7b) ~ Is that correct?

**Comment by YuchenLiu98** [comment]
It is correct. After training, you should directly apply the model for inference without converting

**Comment by rover5056** [comment]
> It is correct. After training, you should directly apply the model for inference without converting

And I also tested stage 1 's output , it generate image's caption correctly . With original vicuna-7b-v1.1 and VIT added pretrain-mm-projector.

But after stage 2 's fine tune,  it becomes garbled...

**Comment by rover5056** [comment]
I solved the problem. Now I can exactly align the model outcome.

It happens when saving model.. Torch' s fsdp may copy weights to cuda,but it does not have enough space so it does not save some layers correctly.

see more details in [pytorch repo](https://github.com/pytorch/pytorch/issues/98823)

**Comment by findalexli** [comment]
How did you resolve the issue?

**Comment by dydxdt** [comment]
> 
Hi, I want to ask that you mean your original bad generation is caused by FSDP?  I met this problem before and changed it to state_dict[fqn] = state_dict[fqn].cpu().clone().detach(). 
But now when I test my finetuned model, the output is still very strange,  for example,""1st 1st 1st 1st 1st 1st 1st 1st 1st 1st ....", "1other 2others 3others\n### 1other 2others 3others\n### 4others 5others\n### 6others 7others\n###". And this happens when I test on my train data. 
I use v0 model. Do you know how to fix this? Thank you!
